### PR TITLE
Refactor Netty reader handler

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Refactor the Netty read handler of worker.

### Why are the changes needed?

The previous implementation creates a state machine per read request, instead of per channel. This implies that if two read requests are sent over the same channel, the worker would possibly use one channel to send data of different files or regions. This can lead to data corruption.

This PR proposes to use a state machine per channel, and handles channel events throughout the whole lifecycle of the channel. Things like a faulty client sending a second request over the same channel before the first request is completed, is handled gracefully with a client error.

The state transitions look like the following:
graph(3)

### Does this PR introduce any user facing changes?

No.

### Tests

Tests have been done with basic Alluxio CLI tools, as well as automated PrestoDB and TPC-DS tests.

### CODE REVIEW

File: dora/core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannel.java [hunk 1]
```diff
@@ -73,7 +73,10 @@ public void readBytes(ByteBuffer outputBuf) {
 
   @Override
   public int readableBytes() {
-    throw new UnsupportedOperationException("DataFileChannel#readableBytes is not implemented.");
+    int lengthInt = (int) mLength;
+    Preconditions.checkArgument(mLength == (long) lengthInt,
+        "size of file %s is %s, cannot be cast to int", mFile, mLength);
```
Comment (huanghua78) [inline dora/core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannel.java:78]: mLength is long. Can it be printed as `%s`?

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadHandler.java [hunk 2]
```diff
@@ -117,21 +103,44 @@ public DataBuffer getDataBuffer(BlockReadRequestContext context, Channel channel
...
+        LOG.warn("Failed to close block reader for block {} with error {}.",
+            mReadRequest.getId(), e.getMessage());
+      }
+    }
+  }
+
+  /**
+   * BlockPackerReaderFactory.
+   */
+  public static class BlockPackerReaderFactory
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadHandler.java:120]: s/BlockPackerReaderFactory/BlockPacketReaderFactory

File: dora/core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannel.java (hunk 1 above)
Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannel.java:78]: Yes, `%s` calls `toString` of the argument.

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadHandler.java (hunk 2 above)
Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadHandler.java:120]: good catch!

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java [hunk 3]
```diff
@@ -30,7 +32,7 @@ public final class BlockReadRequest extends ReadRequest {
    */
   BlockReadRequest(Protocol.ReadRequest request) {
     super(request.getBlockId(), request.getOffset(), request.getOffset() + request.getLength(),
-        request.getChunkSize());
+        downcastChunkSize(request));
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java:35]: I don't understand why this is needed.

```
class ReadRequest {
  private final long mId;
  private final long mStart;
  private final long mEnd;
  private final long mPacketSize;
  private final long mSessionId;

  ReadRequest(long id, long start, long end, long packetSize) {
    mId = id;
    mStart = start;
    mEnd = end;
    mPacketSize = packetSize;
    mSessionId = IdUtils.createSessionId();
  }

```

The constructor expects a long for packetSize.
Why do we need to downcast it to int?

Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java:35]: it is confusing to me that:
Now we are doing "page based reading" in Dora.
But this is called "block read request".
I know the Netty implementation is picked from old version.
But do you think it is a good time to change the terminologies?

Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java:35]: I changed that to `int`. https://github.com/Alluxio/alluxio/pull/17479/files#diff-0e33aad17b8460eb5b842aff737b0c63c189038229f77a5c1a3d657f9c3c4880

A packet is the transfer unit of a read request over the Netty channel. The read request may request a range larger than `int` can represent, but we are never going to have a packet size that exceeds `Integer.MAX_VALUE` which is roughly 2 GiB. On the other hand, we need the packet size to allocate buffers, compute offsets in buffer, etc., and those APIs require an `int` as the length. See https://github.com/Alluxio/alluxio/blob/aa8e38cfa7ebac392f68fad949cc4e344e2573f8/dora/core/server/worker/src/main/java/alluxio/worker/netty/NettyReadHandlerStateMachine.java#L646-L653

Comment (dbw9580) [reply dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java:35]: > do you think it is a good time to change the terminologies

Yes, Jiacheng is removing all block-based APIs from the Dora codebase in #17638 . This PR predates his, so I had to keep the compatibility.

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/FileReadHandler.java [hunk 4]
```diff
@@ -90,63 +77,93 @@ public DoraWorker getWorker() {
...
-        } catch (Exception e) {
-          LOG.warn("Failed to close block reader for block {} with error {}.",
-              context.getRequest(), e.getMessage());
+    public FilePacketReader create(BlockReadRequest readRequest) throws IOException {
+      try {
+        final String fileId =
+            new AlluxioURI(readRequest.getOpenUfsBlockOptions().getUfsPath()).hash();
+        BlockReader reader =
+            mWorker.createFileReader(fileId, readRequest.getStart(),
+                /* positionShort */ false, readRequest.getOpenUfsBlockOptions());
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/netty/FileReadHandler.java:105]: The same thing that is confusing here, again.
We are creating a File Reader, and the result is a Block Reader.

The same question as previous one.

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/ReadRequest.java [hunk 5]
```diff
@@ -23,10 +23,10 @@ class ReadRequest {
   private final long mId;
   private final long mStart;
   private final long mEnd;
-  private final long mPacketSize;
+  private final int mPacketSize;
```
Comment (huanghua78) [inline dora/core/server/worker/src/main/java/alluxio/worker/netty/ReadRequest.java:26]: OK. I found this change now.

Why is this change needed?

File: dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java (hunk 3 above)
Comment (huanghua78) [reply dora/core/server/worker/src/main/java/alluxio/worker/netty/BlockReadRequest.java:35]: OK

**Comment by huanghua78** [comment]
Hello @dbw9580 I suggest to have this nice diagram of control flow or state machine transition to be added in our source code, (in some doc, or in source code)

**Comment by dbw9580** [comment]
Tentatively merge this to unblock testing.
With the current implementations, there are known issues under high concurrent requests, but they will be addressed in follow-up PRs.

aluxio-bot, merge this please.

**Comment by dbw9580** [comment]
alluxio-bot, merge this please.
